File: sampleContract-ts/src/my-private-asset-contract.ts
/*
 * SPDX-License-Identifier: Apache-2.0
 */
import crypto = require('crypto');
import { Context, Contract, Info, Returns, Transaction } from 'fabric-contract-api';
import { MyPrivateAsset } from './my-private-asset';
const myCollectionName: string = 'collectionOne';

@Info({title: 'MyPrivateAssetContract', description: 'My Private Smart Contract' })
export class MyPrivateAssetContract extends Contract {

    @Transaction(false)
    @Returns('boolean')
    public async myPrivateAssetExists(ctx: Context, myPrivateAssetId: string): Promise<boolean> {
        const buffer: Buffer = await ctx.stub.getPrivateData(myCollectionName, myPrivateAssetId);
        return (!!buffer && buffer.length > 0);
    }

    @Transaction()
    public async createMyPrivateAsset(ctx: Context, myPrivateAssetId: string): Promise<void> {
        const exists: boolean = await this.myPrivateAssetExists(ctx, myPrivateAssetId);
        if (exists) {
            throw new Error(`The private asset ${myPrivateAssetId} already exists`);
        }

        const privateAsset: MyPrivateAsset = new MyPrivateAsset();

        const transientData: Map<string, Buffer> = ctx.stub.getTransient();
        if (transientData.size === 0 || !transientData.has('privateValue')) {
            throw new Error('The privateValue key was not specified in transient data. Please try again.');
        }
        privateAsset.privateValue = transientData.get('privateValue').toString('utf8');

        await ctx.stub.putPrivateData(myCollectionName, myPrivateAssetId, Buffer.from(JSON.stringify(privateAsset)));
    }

    @Transaction(false)
    public async readMyPrivateAsset(ctx: Context, myPrivateAssetId: string): Promise<string> {
        const exists: boolean = await this.myPrivateAssetExists(ctx, myPrivateAssetId);
        if (!exists) {
            throw new Error(`The private asset ${myPrivateAssetId} does not exist`);
        }
        let privateDataString: string;
        const privateData: Buffer = await ctx.stub.getPrivateData(myCollectionName, myPrivateAssetId);
        privateDataString = JSON.parse(privateData.toString());
        return privateDataString;
    }

    @Transaction()
    public async updateMyPrivateAsset(ctx: Context, myPrivateAssetId: string): Promise<void> {
        const exists: boolean = await this.myPrivateAssetExists(ctx, myPrivateAssetId);
        if (!exists) {
            throw new Error(`The private asset ${myPrivateAssetId} does not exist`);
        }

        const privateAsset: MyPrivateAsset = new MyPrivateAsset();

        const transientData: Map<string, Buffer> = ctx.stub.getTransient();
        if (transientData.size === 0 || !transientData.has('privateValue')) {
            throw new Error('The privateValue key was not specified in transient data. Please try again.');
        }
        privateAsset.privateValue = transientData.get('privateValue').toString('utf8');

        await ctx.stub.putPrivateData(myCollectionName, myPrivateAssetId, Buffer.from(JSON.stringify(privateAsset)));
    }

    @Transaction()
    public async deleteMyPrivateAsset(ctx: Context, myPrivateAssetId: string): Promise<void> {
        const exists: boolean = await this.myPrivateAssetExists(ctx, myPrivateAssetId);
        if (!exists) {
            throw new Error(`The private asset ${myPrivateAssetId} does not exist`);
        }
        await ctx.stub.deletePrivateData(myCollectionName, myPrivateAssetId);
    }

    @Transaction(false)
    public async verifyMyPrivateAsset(ctx: Context, myPrivateAssetId: string, objectToVerify: MyPrivateAsset): Promise<boolean> {

        // Convert user provided object into a hash
        const hashToVerify: string = crypto.createHash('sha256').update(JSON.stringify(objectToVerify)).digest('hex');
        const pdHashBytes: Buffer = await ctx.stub.getPrivateDataHash(myCollectionName, myPrivateAssetId);
        if (pdHashBytes.length === 0) {
            throw new Error('No private data hash with the key: ' + myPrivateAssetId);
        }

        const actualHash: string = pdHashBytes.toString('hex');

        // Compare the hash calculated (from object provided) and the hash stored on public ledger
        if (hashToVerify === actualHash) {
            return true;
        } else {
            return false;
        }
    }

}

File: sampleContract-java/src/main/java/org/example/MyPrivateAssetContract.java
/*
 * SPDX-License-Identifier: Apache-2.0
 */
package org.example;

import org.hyperledger.fabric.contract.Context;
import org.hyperledger.fabric.contract.ContractInterface;
import org.hyperledger.fabric.contract.annotation.Contract;
import org.hyperledger.fabric.contract.annotation.Default;
import org.hyperledger.fabric.contract.annotation.Transaction;
import org.hyperledger.fabric.contract.annotation.Contact;
import org.hyperledger.fabric.contract.annotation.Info;
import org.hyperledger.fabric.contract.annotation.License;

import java.io.UnsupportedEncodingException;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Map;

@Contract(name = "MyPrivateAssetContract", info = @Info(title = "MyPrivateAsset contract", description = "My Private Smart Contract", version = "0.0.1", license = @License(name = "Apache-2.0", url = ""), contact = @Contact(email = "sampleContract-java@example.com", name = "sampleContract-java", url = "http://sampleContract-java.me")))
@Default
public class MyPrivateAssetContract implements ContractInterface {
    String collection = "collectionOne";

    public MyPrivateAssetContract() {

    }

    @Transaction()
    public boolean myPrivateAssetExists(Context ctx, String myPrivateAssetId) {
        byte[] buffer = ctx.getStub().getPrivateData(collection, myPrivateAssetId);
        return (buffer != null && buffer.length > 0);
    }

    @Transaction()
    public void createMyPrivateAsset(Context ctx, String myPrivateAssetId) throws UnsupportedEncodingException {
        boolean exists = myPrivateAssetExists(ctx, myPrivateAssetId);
        if (exists) {
            throw new RuntimeException("The private asset " + myPrivateAssetId + " already exists");
        }
        MyPrivateAsset privateAsset = new MyPrivateAsset();
        Map<String, byte[]> transientData = ctx.getStub().getTransient();

        if (transientData.size() == 0 | !transientData.containsKey("privateValue")) {
            throw new RuntimeException("The privateValue key was not specified in transient data. Please try again.");
        }
        privateAsset.privateValue = new String(transientData.get("privateValue"), "UTF-8");
        
        ctx.getStub().putPrivateData(collection, myPrivateAssetId, privateAsset.toJSONString().getBytes(StandardCharsets.UTF_8));
    }

    @Transaction()
    public String readMyPrivateAsset(Context ctx, String myPrivateAssetId) throws UnsupportedEncodingException {
        boolean exists = myPrivateAssetExists(ctx, myPrivateAssetId);
        if (!exists) {
            throw new RuntimeException("The private asset " + myPrivateAssetId + " does not exist");
        }

        byte[] privateData = ctx.getStub().getPrivateData(collection, myPrivateAssetId);
        String privateDataString = new String(privateData, "UTF-8");

        return privateDataString;
    }

    @Transaction()
    public void updateMyPrivateAsset(Context ctx, String myPrivateAssetId) throws UnsupportedEncodingException {
        boolean exists = myPrivateAssetExists(ctx, myPrivateAssetId);
        if (!exists) {
            throw new RuntimeException("The private asset " + myPrivateAssetId + " does not exist");
        }
        MyPrivateAsset privateAsset = new MyPrivateAsset();

        Map<String, byte[]> transientData = ctx.getStub().getTransient();

        if (transientData.size() == 0 | !transientData.containsKey("privateValue")) {
            throw new RuntimeException("The privateValue key was not specified in transient data. Please try again.");
        }
        privateAsset.privateValue = new String(transientData.get("privateValue"), "UTF-8");

        ctx.getStub().putPrivateData(collection, myPrivateAssetId, privateAsset.toJSONString().getBytes(StandardCharsets.UTF_8));
    }

    @Transaction()
    public void deleteMyPrivateAsset(Context ctx, String myPrivateAssetId) {

        boolean exists = myPrivateAssetExists(ctx, myPrivateAssetId);
        if (!exists) {
            throw new RuntimeException("The private asset " + myPrivateAssetId + " does not exist");
        }
        ctx.getStub().delPrivateData(collection, myPrivateAssetId);
    }

    @Transaction()
    public boolean verifyMyPrivateAsset(Context ctx, String myPrivateAssetId, MyPrivateAsset objectToVerify) throws NoSuchAlgorithmException {
        
        // Convert user provided object into hash
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        byte[] hashByte = digest.digest(objectToVerify.toJSONString().getBytes(StandardCharsets.UTF_8));

        String hashToVerify = new BigInteger(1, hashByte).toString(16);

        // Get hash stored on the public ledger
        byte[] pdHashBytes = ctx.getStub().getPrivateDataHash(collection, myPrivateAssetId);
        if (pdHashBytes.length == 0) {
            throw new RuntimeException("No private data hash with the key: " + myPrivateAssetId);
        }

        String actualHash = new BigInteger(1, pdHashBytes).toString(16);

        if (hashToVerify.equals(actualHash)) {
            return true;
        } else {
            return false;
        }
    }

}
